packages/SqueakMaps-Core.package/SMASearchManager.class/instance/getRouteFrom.To.By..st

```diff
@@ -1,12 +1,8 @@
 searching
-getRouteFrom: aQueryString To: anotherQueryString By: aSymbol
+getRouteFrom: aStartLocation To: aDestinationLocation By: aSymbol
```

Comment:
Ist es wirklich sinnvoll die Signatur der Methode zu ändern? Wenn ich jetzt eine Route suchen möchte, reicht es nicht mehr einfach nur zwei Ortsnamen einzugeben, sondern ich muss ein vollständiges Location Objekt hineingeben, was ich mir erst mal vorher erstellen muss. Zumindest war das ursprünglich genau das Reasoning was wir beim Search Manager hatten, dass man möglichst wenige Informationen vorher selbst erstellen muss und er den ganzen Kram für uns übernimmt.

Ich sehe auf jeden Fall warum ihr das brauchtet, nämlich für die zwei Pins die dann angezeigt werden sollen. Aber dadurch explodiert halt die `searchDirections` Methode, die wir ursprünglich durch den Search Manager verschlanken wollten.

Das überschneidet sich ja so ein bisschen mit dem Error Handling von Ole und mir. Kann man vielleicht nur im Fehlerfall die zwei gesuchten Orte an das Window zurückgeben, wobei der Fehler abgefangen wird und damit eindeutig klar ist, dass nur zwei Orte zurückkommen?

Reply:
Ja, uns war bewusst, dass das wieder etwas Verantwortung in das Window zieht. Wir hatten dann auch über deine vorgeschlagene Version nachgedacht, abhängig vom Fehler etwas anderes als die Route zurückzugeben. Da hatten wir aber keine Idee wie man das schlank und leicht verständlich umsetzen könnte (und das Window müsste dann trotzdem noch eine Fallunterscheidung der beiden Rückgabetypen machen). Deswegen haben wir dann gesagt wir sacrificen die Methodenlänge im Window, um den Searchmanager nicht komplizierter zu machen. Wenn Du eine konkrete Idee hast wie man das cool umsetzen könnte ohne viel Overhead zu erzeugen können wir ja aber auch nochmal reinschauen 👍

Reply:
J

> Achso, genau, falls Ihr da natürlich jetzt sowieso ein Fehlerobjekt werft, könnte man da die zwei Locations mit drauf speichern und die sich da dann rausholen.

Ja genau, an so etwas habe ich auch gedacht. Bevor wir das implementieren nochmal kurz was dafür gemacht werden müsste: Wir werfen den `RouteTooLongError` oder den `UnreachableDestinationError` in der Routing API. Die Routing API bekommt ja keine Locations in die `getRouteFrom` Methode, sondern nur Geographische Koordinaten, also kann auch auf den Errors nicht direkt die Locations setzen.

Der Call an die `getRouteFrom` Methode kommt ja vom Serach Manager, deshalb müssten wir dort die Errors abfangen uns könnten erst dort die Locations setzen und signaln dann den Error einfach nochmal.

Ist das eine halbwegs cleane Lösung?

Reply:
Mh ok, ja ich denke das wäre clean, aber ich bin mir nicht sicher, ob es der Verständlichkeit des Projekts zuträgt. Ich hab eher den Eindruck, dass wir hier over-engineeren. Ursprünglich war ja das Ziel, die Methode auf dem Window klein zu halten, weswegen der Search Manager eingeführt wurde, der nun Search-Terms erwartet. Jetzt haben sich aber die Requirements geändert (wir brauchen die tatsächlichen Locations im Window). Wir könnten sie einfach im Window erstellen, wodurch ca. 3-4 (?) Zeilen mehr dazu kommen und der Search-Manager direkt Locations erwartet. Oder wir lassen den Search Manager so, müssen dann aber im Window auf spezifische Fehler prüfen (+ 1-2 Zeilen (?)). Zusätzlich dazu muss der Fehler aber im Search Manager zwischendurch abgefangen und verändert werden (+ 3-4 Zeilen (?)), wodurch man sich bei der Fehlerklasse jetzt nicht mehr sicher sein kann, ob denn Locations gesetzt sind oder nicht (abhängig da wo man ihn abfängt). Der letzte Punkt wäre, dass man diese Funktionalität dann für jeden Routing Fehler einzeln implementieren muss, oder? (Falls man nicht auch noch Vererbung einführen will :D) Aktuell haben wir aber die Möglichkeit, bei jedem Routing-Fehler trotzdem die Start- und End-Location anzuzeigen, direkt out-of-the-box.
All das zusammen betrachtet wirkt es für mich so als wäre eine etwas längere Window-Methode die bessere Wahl, zumindest was die Komplexität und Verständlichkeit der Codebase für neue Entwickler angeht. Alternativ könnte man ja aber auch überlegen, andere Punkte (wie Checken der Inputs) auszulagern.

Reply:
Mhh...ich finds zwar immer noch nicht nice, das Interface des Search Managers so zu ändern, eben weil er jetzt im Prinzip nur noch Calls weiterleitet (das was ja am Anfang deine Befürchtung war) und man nun eine komplette Location reingeben muss anstatt einfach nur dem Namen.

Mir fällt jetzt aber spontan nichts besseres ein. Höchstens man ändert das Interface der 'getRouteFromToBy' Methode auf der Routing API und gibt statt Koordinaten ganze Locations rein. Das hatten wir ursprünglich nur so gemacht weil die API an sich nur Koordinaten nimmt, aber die Nutzung von Locations würde meiner Meinung nach in dem Package auch Sinn machen.

Dann müsste man das nur trotzdem irgendwie in die Error Message reinkriegen und würde im Window auch das Anzeigen der zwei Locations für mehrere Fehler mehrmals aufrufen müssen.

Macht wie ihr denkt, entweder ihr lagert die Input Test Calls noch aus oder was anderes. Hauptsache wir kriegen es bald gemergt.

Reply:
Haben uns jetzt geeinigt, das erst mal so zu übernehmen und den SearchManager nochmal zu einem späteren Zeitpunkt zu überdenken.